// craco/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  overrideEslint: true,
  overrideTypeScript: true,
  overrideBabel: true
};
exports.overrideEslint = overrideEslint;
exports.overrideTypeScript = overrideTypeScript;
exports.overrideBabel = overrideBabel;

var _babel = require("@craco/craco/lib/features/webpack/babel");

var _eslint = require("@craco/craco/lib/features/webpack/eslint");

var _typescript = require("@craco/craco/lib/features/webpack/typescript");

var _craco = require("@craco/craco");

Object.keys(_craco).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _craco[key];
    }
  });
});

function overrideEslint(eslintOptions, webpackConfig) {
  return (0, _eslint.overrideEslint)({
    eslint: eslintOptions
  }, webpackConfig);
}

function overrideTypeScript(typescriptOptions, webpackConfig) {
  return (0, _typescript.overrideTypeScript)({
    typescript: typescriptOptions
  }, webpackConfig);
}

function overrideBabel(babelOptions, webpackConfig) {
  return (0, _babel.overrideBabel)({
    babel: babelOptions
  }, webpackConfig);
}
